controller/aks-cluster-config-handler.go

```diff
@@ -498,386 +479,132 @@ func BuildUpstreamClusterState(ctx context.Context, secretsCache wranglerv1.Secr
-	client.Authorizer = authorizer
-
-	return &client, nil
-}
-
-// createOrUpdateCluster creates a new managed Kubernetes cluster
-func createOrUpdateCluster(ctx context.Context, secretsCache wranglerv1.SecretCache, clusterClient *containersv211.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (c containersv211.ManagedCluster, err error) {
-
-	dnsPrefix := spec.DNSPrefix
-	if dnsPrefix == nil {
-		dnsPrefix = to.StringPtr(spec.ClusterName)
-	}
-
-	tags := make(map[string]*string)
-	for key, val := range spec.Tags {
-		if val != "" {
-			tags[key] = to.StringPtr(val)
+	// check tags for update
+	if spec.Tags != nil {
+		if !reflect.DeepEqual(spec.Tags, upstreamSpec.Tags) {
+			logrus.Infof("Updating tags for cluster [%s]", spec.ClusterName)
+			tags := containerservice.TagsObject{
+				Tags: *to.StringMapPtr(spec.Tags),
+			}
+			_, err = resourceClusterClient.UpdateTags(ctx, spec.ResourceGroup, spec.ClusterName, tags)
+			if err != nil {
+				return spec, err
+			}
 		}
 	}
-	displayName := spec.ClusterName
-	if displayName == "" {
-		displayName = spec.ClusterName
-	}
-	tags["displayName"] = to.StringPtr(displayName)
-
-	kubernetesVersion := spec.KubernetesVersion
-
-	var vmNetSubnetID *string
-	networkProfile := &containersv211.NetworkProfile{}
-	if hasCustomVirtualNetwork(spec) {
-		virtualNetworkResourceGroup := spec.ResourceGroup
-
-		//if virtual network resource group is set, use it, otherwise assume it is the same as the cluster
-		if spec.VirtualNetworkResourceGroup != nil {
-			virtualNetworkResourceGroup = *spec.VirtualNetworkResourceGroup
-		}
-
-		vmNetSubnetID = to.StringPtr(fmt.Sprintf(
-			"/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v/subnets/%v",
-			spec.SubscriptionID,
-			virtualNetworkResourceGroup,
-			spec.VirtualNetwork,
-			spec.Subnet,
-		))
-
-		networkProfile.DNSServiceIP = spec.NetworkDNSServiceIP
-		networkProfile.DockerBridgeCidr = spec.NetworkDockerBridgeCIDR
-		networkProfile.ServiceCidr = spec.NetworkServiceCIDR
-
-		if spec.NetworkPlugin != nil {
-			networkProfile.NetworkPlugin = containersv211.Kubenet
-		} else {
-			networkProfile.NetworkPlugin = containersv211.NetworkPlugin(*spec.NetworkPlugin)
-		}
-
-		// if network plugin is 'kubenet', set PodCIDR
-		if networkProfile.NetworkPlugin == containersv211.Azure {
-			networkProfile.PodCidr = spec.NetworkPodCIDR
-		}
 
-		if spec.LoadBalancerSKU != nil {
-			loadBalancerSku := containersv211.LoadBalancerSku(*spec.LoadBalancerSKU)
-			networkProfile.LoadBalancerSku = loadBalancerSku
-		}
-
-		if spec.NetworkPolicy != nil {
-			networkProfile.NetworkPolicy = containersv211.NetworkPolicy(*spec.NetworkPolicy)
-		}
+	agentPoolClient, err := aks.NewAgentPoolClient(credentials)
+	if err != nil {
+		return spec, err
 	}
 
-	var agentPoolProfiles []containersv211.ManagedClusterAgentPoolProfile
-	if hasAgentPoolProfile(spec) {
+	needsUpdate := false
+	// check NodePools for update
+	if spec.NodePools != nil {
```

Comment:
Checking the node pools for updates first means we can't update the cluster kubernetes version and a node pool kubernetes version at the same time. If the cluster version was updated first then the node pool versions could follow and there wouldn't be an error.

Reply:
We have to check what is recommended upgrade procedure for AKS clusters by Azure doc

Reply:
Have this been checked?